Comparison

etienneschmitt opened the pull request:
Ajout de benchs permettant de comparer cgogn à d'autres libs : 
-cgogn1
-cgal LCC
-OpenMesh
-SurfaceMesh

Ce sont des tests qui proviennent du package LCC de CGAL. Je ne sais pas s'ils sont tous pertinents, j'attends vos retours. Je pense qu'on devrait y ajouter ceux développés de notre côté.
Ces benchmarks ne couvrent pour l'instant que les 2-cartes. Le reste suivra dans un second temps.

etienneschmitt committed:
updated google benchmark.

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Added some benchmark to compare CGoGN_2 to other libraries.
For the moment those libraries are tested :
-cgal LLC
-SurfaceMesh
-OpenMesh
-CGoGN_1
and obviously CGoGN_2.
The test have been found in the git repository of CGAL, in the LinearCellComplex module.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

etienneschmitt committed:
Removed unused cmake modules.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

etienneschmitt committed:
MapBaseData : move some thread related method to the cpp file.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

etienneschmitt committed:
ChunkArrayContainer : added remove_marker_attributes method.

Added a call to invalidate_external_refs()	 in the remove_chunk_arrays() method.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

etienneschmitt committed:
MapBase: Calling remove_marker_attributes() in the clear_and_remove_attributes() method.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

etienneschmitt committed:
Fixed compilation with older version of google benchmark.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

etienneschmitt committed:
Fixed compilation.

Signed-off-by: Étienne Schmitt <etienne@schmitt.wf>

pierrekraemer commented on cgogn/core/cmap/map_base_data.cpp:
```diff
@@ -42,10 +42,9 @@ MapBaseData::MapBaseData()
-
 	// register the map in the vector of instances
-	instances_->push_back(this);
+	if (std::find(instances_->begin(), instances_->end(), this) == instances_->end())
```
Qu'est-ce qui justifie le passage de assert en if ici ? C'est qqch qui est censé pouvoir arriver qu'on passe dans le constructeur d'une carte déjà présente dans le vector d'instances ?

etienneschmitt replied:
Non. On pourrait même virer le if.

pierrekraemer commented on cgogn/core/cmap/map_base.h:
```diff
@@ -109,8 +109,10 @@ class MapBase : public MapBaseData
+		this->topology_.remove_marker_attributes();
 		this->topology_.clear_chunk_arrays();
 
+		boundary_marker_ = topology_.add_marker_attribute();
```
Je me souviens que l'on avait dit que lors d'un clear_and_remove_attributes, on conservait les markers. D'ailleurs, un peu plus bas (nouvelle ligne 131), on les nettoie. Si ils sont supprimés, pas besoin de les nettoyer. Je suis d'avis de conserver les markers.

etienneschmitt replied:
Et bien figure-toi que dans les benchmarks, avant de recourir à des unique_ptr et à faire des delete/new sur la map entre chaque itération, j'utilisais clear_and_remove_attributes. Et je me suis rendu compte que les performances empiraient d'une itération à l'autre. Cette modification résout ce problème de performance.

etienneschmitt replied:
Et comme tu le dis les zéro markers restant sont nettoyés après la suppression. Il ne se passe donc rien.

pierrekraemer replied:
Plutôt que de supprimer les markers en même temps que les attributs, je serais d'avis de creuser un peu la raison de cette baisse de performances.
Pourrait-on traquer la quantité de mark_attribute dans les containers ? Normalement, chaque Dart/CellMarker récupère et libère un mark_attribute dans un pool maintenu pour chaque thread par la carte (mark_attributes_ et mark_attributes_topology_). Il faudrait voir si il y a un problème là dedans.

etienneschmitt replied:
Oui je vais creuser ça.

untereiner commented:
Ne serait-il pas judicieux de mettre tous ces bench dans un dépôt à côté de celui du code de la lib ? Sinon, j'ai donné mes benchs à Guillaume Damiand, qui les a mis dans cgal, j'ai les mêmes pour les volumes.

etienneschmitt commented:
Je dirais que le dossier benchmark fait très bien l'affaire pour l'instant. OK pour les benchs, donc il y a de fortes chances que tu ais écrit les benchs que j'ai rajoutés dans cette PR ;-)

etienneschmitt commented:
J'ai aussi vu ceux pour les volumes, je m'y attaquerai quand j'aurai un peu de temps.

untereiner commented:
SI tu veux y rajouter des comparaisons avec d'autres libs et donc des dépendances à ces libs, je considère que ce dossier n'a plus sa place dans le projet "code d'un noyau de modeleur" mais que c'est un projet à part entier dont cgogn2 est une dépendance au même titre que les autres libs

etienneschmitt committed:
CellMarker : removed unnecessary constructors.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
ChunkArrayContainer : removed the remove_marker_attributes() method.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
ChunkArrayContainer : constness.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
MapBase : updated the clear_and_remove_attributes() method.
New behaviour : the markers are not deleted but instead they are cleared.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
MapBaseData : added a swap_chunk_array_container().
This method was first in the MapBuilder classes but it was incomplete.
Now it should correctly update the internal data structure of the map.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
ChunkArray : some cleaning and optimization.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
back to assert.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt commented:
Voilà j'ai un peu cherché et je pense avoir reglé le problème d'une façon plus maline.

etienneschmitt commented:
Concernant les dépendances tu as dû mal regarder, il n'y en a aucune.

etienneschmitt committed:
ChunkArray : small fix.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Removed wrong const qualifier on ChunkArrayBool return type.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

untereiner reviewed (changes requested):
Toute la partie comparaison est hors du champs d'application du projet qui est: une lib de modélisation. Cela fait grossir la base de code inutilement. Il faudrait garder les commits qui améliorent les performances de la lib et déplacer les autres dans un autre dépôt.  
Edit: suis-je bête, un find_package n'est pas une dépendance

pierrekraemer commented on cgogn/core/cmap/map_base.h:
```diff
@@ -111,9 +111,6 @@ class MapBase : public MapBaseData
-
 		for (auto& att : this->attributes_)
 			att.remove_chunk_arrays();
 
```
Ici il manque un nettoyage des markers de topology :
```c++
std::lock_guard<std::mutex> lock(this->mark_attributes_topology_mutex_);;
for (std::vector<ChunkArrayBool*>& mark_att_topology_thread_i : mark_attributes_topology_)
  for(ChunkArrayBool* cab : mark_att_topology_thread_i)
    cab->clear();
```
Le code qui était là avant me semble vraiment très bizarre.. Il était écrit :
```c++
for (auto& mark_att_topo : this->mark_attributes_topology_)
  mark_att_topo.clear();
```
Ca revient à appeler clear sur le vector de chaque thread qui contient les ChunkArrayBool*, alors que ce que l'on souhaite faire, c'est nettoyer les markers (appeler la méthode clear des ChunkArrayBool). Le code qui était écrit revenait à oublier les ChunkArrayBool existant ce qui fait qu'ils s'entassaient dans les containers après chaque clear_and_remove_attributes.

etienneschmitt replied:
Oui c'est ça il y avait une grosse confusion ici.

etienneschmitt replied:
Voilà la partie nettoyage des markers de topology est ajouté. Pour les parcourir j'ai choisi d'utiliser la méthode marker_arrays() que j'ai mise dans ChunkArrayContainer.

etienneschmitt committed:
MapBase : cleaning topology markers in the clear_and_remove_attributes() method.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt commented:
En regardant de plus près @untereiner , tu verras les find_package que j'ai mis n'ont pas le mot clef REQUIRED. Les benchs compilent très bien avec juste cgogn2. Bien sûr dans ce cas la comparaison sera limitée.
Ensuite concernant l'idée de mettre ces benchs ailleurs, je réponds que non, ils sont à leur place ici.
Cordialement.

untereiner commented:
Oui, c'est une dépendance, optionnelle, mais une dépendance quand même. Tout ceci sans autre argument que "ils sont à leur place ici". Soit, je m'incline.

etienneschmitt committed:
Added comparison to CGAL/Polyhedron.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
OpenVolumeMesh files ("*.ovm) IO.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
map3 comparison (early stage).

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

pierrekraemer commented on cgogn/core/cmap/map_base.h:
```diff
@@ -122,11 +125,15 @@ class MapBase : public MapBaseData
+				std::lock_guard<std::mutex> lock(this->mark_attributes_topology_mutex_);
+				for (ChunkArrayBool* cab : this->topology_.marker_arrays())
+					cab->clear();
+			}
```
En revanche, ce parcours des attributs de marker du container de topology n'a rien à faire dans la boucle sur les orbites. Rien ne dépend de i ici et il ne faut faire ça qu'une seule fois.

etienneschmitt committed:
Revert "OpenVolumeMesh files ("*.ovm) IO."

This reverts commit 4c9d4ac6f8053f6358847e883dad136456d056c7.

etienneschmitt committed:
Fixed CMake problem.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Fixed MapBase::clear_and_remove_attributes() method.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Fixed compilation of map2 and map3 comparison with the latest version of google benchmark.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

pierrekraemer commented:
C'est quoi le fichier "comparison.patch" ?

etienneschmitt commented:
Un fichier qui n'a rien à faire ici. Je fais un dernier commit en essayant de nettoyer les nombreuses coquilles de cette PR et ce sera bon de mon côté.

etienneschmitt committed:
Removed trash file.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Added tetmesh I/O.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Added an OpenVolumeMesh sample file (aneurysm_3D.ovm).

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Added LCC and cgogn2 performances to the map3 comparison benchmark.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Fixed compilation with cgal < 4.9.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

pierrekraemer commented:
Bon, si on a fini avec la correction des problèmes de gestion d'attributs de marquage, je donne mon avis sur l'autre discussion qui anime cette PR.
On a depuis longtemps un dossier "benchmarks" dans lequel il y a du code pour mesurer des temps d'exécution. Qu'il y ait là dedans du code qui permette de mesurer des temps d'exécution pour des algos communs sur des libs autre que cgogn dans le cas où ces dernières sont disponibles sur la machine, cela ne me dérange pas plus que ça.
Si on ne veut que le code du coeur de la lib sur ce dépôt, alors on peut supprimer complètement le dossier "benchmarks", et en poussant un peu dans ce sens, on pourrait aussi déplacer les "tests" et les "examples" dans un autre dépôt. Cela ne me semble pas souhaitable.
En revanche, la compilation des benchs peut, si l'utilisateur le souhaite, ne pas être faite du tout. A ce sujet, je serais d'avis que dans le CMakeLists racine, les options CGOGN_BUILD_BENCHS, CGOGN_BUILD_EXAMPLES et CGOGN_BUILD_TESTS soient à OFF. Ainsi, par défaut, on ne compile que la lib. En tant que développeur, on souhaitera certainement compiler au moins les tests (et aussi probablement les exemples), mais cela reste facile à configurer.

etienneschmitt committed:
Changed default value for somes options:
-CGOGN_BUILD_TESTS is OFF
-CGOGN_BUILD_EXAMPLES is OFF
-CGOGN_BUILD_BENCHS is OFF.

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt commented:
D'accord avec tout ça. J'ai rajouté un commit pour l'appliquer et j'ai mis l'intégration continue à jour pour qu'elle active tout le temps les tests et les examples.

untereiner commented:
Et pourtant, ce découpage c'est la bonne pratique employée par une bonne partie des projets ici ou ailleurs. Je peux ajouter un paquet de liens pour le démontrer si nécessaire. Maintenant, si la majorité ici ne souhaite pas sortir des années 2000 sans autre argument que "c'est plus pratique pour nous" je m'incline. Mes contributions étant désormais assez limitées je ne voudrais pas imposer ma vision. 🙈

hananfirdaws commented:
Bonsoir
Je me nomme Hakem et je me prénomme Hanen. Je suis étudiante à l'université l’Arbi Ben M’hidi Oum El Bouaghi (Algerie). Je suis en Master 2 informatique option Architecture distribuée. 
Le thème de mon mémoire de fin d'étude s’intitule : Utilisation de la librairie CGOGN pour le développement d’un modèle topologique parallèle. 
J’ai besoin de votre aide et j’ai plusieurs questions à vous poser pourriez-vous me répondre si vous acceptez de m’aider 
Dans l’attente de votre réponce svp
Merci et bien cordialement.

pierrekraemer commented:
Disons que pour le moment, ça me semble effectivement plus pratique comme ça. Comme tu l'avais justement fait remarquer toi-même, CGoGN n'a pas actuellement une base de développeurs / utilisateurs particulièrement importante.. Et je ne pense pas que cela soit dû au problème dont on est en train de parler. Quoi qu'il en soit, si on a un jour de nombreux utilisateurs qui souhaitent ce changement, cela ne sera pas difficile à réaliser.

untereiner commented:
Je ne vois pas de corrélation directe entre le nombre de développeurs/utilisateurs et le recentrage de l'objectif de ce dépôt de code. Par contre, ce dont je suis convaincu c'est que ce changement sera encore plus difficile à effectuer s'il y a plus de personnes impliquées (j'ai le cas sous les yeux avec sofa). T'es le boss, on peut donc clore le sujet et accepter toutes ces modifications.

etienneschmitt commented:
Moi c'est vraiment sur le fond que je ne suis pas d'accord. Le projet "CGoGN_2" intègre non seulement la librairie, mais des petites applications (qui ont valeur d'exemples), des tests, et des benchmarks. Qu'y trouves-tu de choquant ? Quelle "bonne pratique" dont tu parles violons-nous ? Est-ce que tu as protesté quand on a décidé de faire des modules avec à chaque fois un dossier "tests" et "examples" ?
Je trouve que tu fais beaucoup de zèle pour ce qui me semble être un point de détail. Pour le cas des benchmark, le tout est dans un dossier "benchmark", à côté du dossier "cgogn". Tout le monde comprend sa raison d'être ici: permettre de mesurer les performances de cgogn (seul ou face à d'autres libs). Ce genre d'outil est directement lié au projet ce qui justifie sa présence sur le dépôt.
De ton côté tu as beaucoup critiqué mais dans aucun de tes messages je ne vois la moindre justification de ton point de vue. Tu parles de bonnes pratiques sans les décrire et d'innombrables projets sans en citer un seul...

untereiner commented:
😆 Toujours la même histoire, même après ton départ de mimesis. Pour les libs, et juste pour rester dans la communauté, au choix: cgal (doc) / sofa (doc) / libigl (doc/tests/examples). J'étais contre les dossiers tests et examples mais je n'ai malheureusement pas d'enregistrement 💩 .... Nous ne serons jamais d'accord donc pour ma part ça s'arrête ici.
